Make location provider optional

Guardiola31337 opened the pull request:
Per [mapbox-gl-native/#9403](https://github.com/mapbox/mapbox-gl-native/issues/9403) we're removing location dependencies on the map SDK. They'll be pulled in from here (MAS), so this PR addresses that.

- [x] Bundle default `LocationEngine` implementations for Android core, LOST, and Google Play Services backends
- [x] Add LOST and Google Play Services to `build.gradle` as optional dependencies
- [x] At runtime, detect whether LOST or Google Play Services are available
- [x] Let the developer set a priority preference for location provider backend (i.e. if more than one provider is available, be able to set one as preferred)

Am I missing something? @mapbox/android 

👀 @zugaldia @tobrun for review

zugaldia commented:
@Guardiola31337 So excited to see this happening.

tobrun commented on mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/LocationSourceProvider.java:
```diff
@@ -0,0 +1,27 @@
+  private void initLocationSourceDictionary(Context context) {
+    locationSourceDictionary = new HashMap<>(3);
+    locationSourceDictionary.put("Google Play Services", GoogleLocationEngine.getLocationEngine(context));
+    locationSourceDictionary.put("Lost", LostLocationEngine.getLocationEngine(context));
```
above doesn't seem to validate if it's in the classpath or not

Guardiola31337 commented:
- [ ] `GoogleLocationEngine` is not working properly (not receiving location changes)
- [ ] Add `LocationEngineChainSupplier` remaining tests
- [ ] Add `LocationEngineProvider` tests

Guardiola31337 commented:
> Let the developer set a priority preference for location provider backend (i.e. if more than one provider is available, be able to set one as preferred)

Usage:

```java
LocationEngineProvider locationEngineProvider = new LocationEngineProvider(this);
locationEngine = locationEngineProvider.obtainAvailableLocationEngines()
  .get(LocationEngineProvider.GOOGLE_PLAY_SERVICES);
if (locationEngine == null) {
  locationEngine = locationEngineProvider.obtainAvailableLocationEngines()
    .get(LocationEngineProvider.ANDROID);
}
```
What do you think @zugaldia?

> `GoogleLocationEngine` is not working properly (not receiving location changes)

We should revisit the provider strategy. I run some tests and the problem was that when you set the priority to `HIGH_ACCURACY` we're using the `GPS_PROVIDER` and this provider takes some time to start receiving location updates (indoors takes forever). I tried to change the priority to `LOW_POWER` or `BALANCED_POWER_ACCURACY` and it worked. Maybe we could add timeout timers so if we don't get any updates on a specific provider and timer elapses use a different provider. Another option to explore is using [`Criteria`](https://developer.android.com/reference/android/location/Criteria.html) as [previously suggested](https://github.com/mapbox/mapbox-java/blob/master/mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/AndroidLocationEngine.java#L96).

> Add `LocationEngineChainSupplier` remaining tests

Not necessary anymore. `LocationEngineChainSupplier` was removed.

> Add `LocationEngineProvider` tests

`LocationEngine` classes are full of static methods and singletons which make difficult adding tests. I guess this could wait until we rewrite those classes.

Guardiola31337 commented:
https://github.com/mapbox/mapbox-java/pull/502/commits/f4f74e3279c021e533c4b32b7f1146f8fd249a2f adds/fixes new APIs to get `LocationEngine`s through `LocationEngineProvider`:

- `obtainBestLocationEngineAvailable()` which returns the best `LocationEngine` available:
  - Return: `GoogleLocationEngine`, `LostLocationEngine` or `GoogleLocationEngine`
  - Criteria: `GOOGLE_PLAY_SERVICES` > `LOST` > `ANDROID`
  - `@NonNull` ~> If `GoogleLocationEngine` and `LostLocationEngine` aren't available `GoogleLocationEngine` is always returned
  - Usage:

```java
LocationEngineProvider locationEngineProvider = new LocationEngineProvider(context);
locationEngine = locationEngineProvider.obtainBestLocationEngineAvailable();
```

- `obtainLocationEngineBy(LocationEngine.Type type)` which returns the `LocationEngine` by type or `null` if not available:
  - Options: `LocationEngine.Type` types (`GOOGLE_PLAY_SERVICES`, `LOST`, `ANDROID`, `MOCK`)
  - `@Nullable` ~> If the location engine you're asking for is not available a `null` is returned
  - Usage:

```java
LocationEngineProvider locationEngineProvider = new LocationEngineProvider(this);
locationEngine = locationEngineProvider.obtainLocationEngineBy(LocationEngine.Type.GOOGLE_PLAY_SERVICES);
if (locationEngine == null) {
  locationEngine = locationEngineProvider.obtainBestLocationEngineAvailable();
}
```

Or directly requesting `ANDROID` location engine which won't be `null` (always available):

```java
LocationEngineProvider locationEngineProvider = new LocationEngineProvider(this);
locationEngine = locationEngineProvider.obtainLocationEngineBy(LocationEngine.Type.GOOGLE_PLAY_SERVICES);
if (locationEngine == null) {
  locationEngine = locationEngineProvider.obtainLocationEngineBy(LocationEngine.Type.ANDROID);
}
```

That ☝️ allows you to:

```java
LocationEngineProvider locationEngineProvider = new LocationEngineProvider(this);
locationEngine = locationEngineProvider.obtainLocationEngineBy(LocationEngine.Type.ANDROID);
```

Could be useful/interesting in some use cases.

---

- [ ] Add javadoc to public methods and classes (documentation)

Apart from ☝️ this PR is ready for review.

👀 @zugaldia @tobrun

zugaldia commented on mapbox/app/build.gradle:
```diff
@@ -59,6 +59,11 @@ dependencies {
 
+    // LOST
+    compile(rootProject.ext.dep.lost) {
+        exclude group: 'com.google.guava'
```
If we upgrade the dependency to the latest LOST that Guava exclusion is no longer necessary.

Guardiola31337 replied:
I tested the sample app with lost `3.0.3` (atm latest version) and I found that it wasn't working properly so I'd rather keep lost `1.1.1` until https://github.com/mapzen/lost/issues/241 gets fixed.

zugaldia replied:
@Guardiola31337 I'd actually argue the opposite. With this PR the developer now has the freedom to switch location providers if they don't meet their needs. I think that all that https://github.com/mapzen/lost/issues/241 shows is that LOST fails under certain scenarios. I'd still upgrade to the latest version of LOST and leave it up to the developer decide if that scenario is a blocker for them.

zugaldia commented on mapbox/app/src/main/java/com/mapbox/services/android/testapp/geocoding/GeocodingServiceActivity.java:
```diff
@@ -122,10 +122,14 @@ private void updateValuesFromBundle(Bundle savedInstanceState) {
-   * Builds Android location engine
+   * Builds location engine
    */
-  private synchronized void buildAndroidLocationEngine() {
```
Why is `synchronized` no longer needed?

Guardiola31337 replied:
@zugaldia 
Don't like to answering with a question but I'm pretty sure you have more context:
Why was `synchronized in the first place? Why `getLocationEngine()` in `AndroidLocationEngine`, `LostLocationEngine` and `GoogleLocationEngine` need to be `synchronized`?

BTW, I've tested it and I didn't find any issues so apparently is working as expected.

zugaldia replied:
@Guardiola31337 For thread safety, did your tests include that scenario (i.e. multi-threading)?

zugaldia replied:
@Guardiola31337 just seeing this is an `Activity`, I though this was part of the library, thanks for pointing that out. Nevermind.

zugaldia commented on mapbox/libandroid-telemetry/build.gradle:
```diff
@@ -32,11 +32,21 @@ dependencies {
+
+    // LOST
+    provided(rootProject.ext.dep.lost) {
+        exclude group: 'com.google.guava'
```
Same here: let's update to the latest version. If the latest version is an issue to the developer, now they have a mechanism to switch to a different provider.

Guardiola31337 replied:
👀 https://github.com/mapbox/mapbox-java/pull/502#discussion_r142348095

zugaldia commented on mapbox/app/src/main/res/values/strings.xml:
```diff
@@ -90,7 +90,7 @@
         <item>Android</item>
         <item>Lost</item>
-        <item>Google Play Services</item>
+        <item>Google_Play_Services</item>
```
Was this necessary to match the new `Type` enum? If so, 👍 .

zugaldia commented:
@tobrun @cammace It'd be great if you could give this PR a pass before merging to make sure there're no concerns on the Maps/Navigation side.

Guardiola31337 commented:
https://github.com/mapbox/mapbox-java/pull/502/commits/2f856d60d9b6dbb3fa9f85c9005df4c09f1edc66 and https://github.com/mapbox/mapbox-java/pull/502/commits/2ec7c2dd6c0b0b4dc11f57f9f140f518ce8b2ce2 addressed your comments @zugaldia 

Ready for a second round of 👀

cammace commented on mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/GoogleLocationEngine.java:
```diff
@@ -12,16 +12,13 @@
  * Sample LocationEngine using Google Play Services
  */
-public class GoogleLocationEngine extends LocationEngine implements
+class GoogleLocationEngine extends LocationEngine implements
```
This still needs to be public inorder not to break the API

cammace commented on mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/GoogleLocationEngine.java:
```diff
@@ -31,7 +28,7 @@
   private GoogleApiClient googleApiClient;
 
-  public GoogleLocationEngine(Context context) {
+  private GoogleLocationEngine(Context context) {
```
needs to remain public to not break API

cammace commented on mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/LostLocationEngine.java:
```diff
@@ -17,20 +15,25 @@
  * Sample LocationEngine using the Open Source Lost library
  */
-public class LostLocationEngine extends LocationEngine implements LocationListener {
+class LostLocationEngine extends LocationEngine implements
```
needs to remain public to not break API

cammace commented on mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/location/LostLocationEngine.java:
```diff
@@ -17,20 +15,25 @@
   private LostApiClient lostApiClient;
 
-  public LostLocationEngine(Context context) {
+  private LostLocationEngine(Context context) {
```
needs to remain public to not break API

Guardiola31337 commented:
@cammace 
> Looks like this won't affect Navigation much, we do fall back on the LOST location engine currently in case a user doesn't pass in a location engine. This change I believe will require us to bring the LOST dependency directly into the nav sdk?

Lost dependency is not needed anymore we could use:
```java
LocationEngineProvider locationEngineProvider = new LocationEngineProvider(context);
locationEngine = locationEngineProvider.obtainBestLocationEngineAvailable();
```
which will return the best location engine available.
